File: answers/leetcode/Letter Combinations of a Phone Number/Letter Combinations of a Phone Number.java
class Solution {
    public List<String> letterCombinations(String digits) {
        final String[] PHONE = new String[]{"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        List<String> combinations = new LinkedList<String>();
        combinations.add("");
        for (int i = 0; i < digits.length(); i++) {
            int num = Character.getNumericValue(digits.charAt(i));
            List<String> newCombinations = new LinkedList<String>();
            for (char c : PHONE[num].toCharArray()) {
                for (String combination : combinations) {
                    newCombinations.add(combination + c);
                }
            }
            combinations = newCombinations;
        }
        if (combinations.size() == 1) {
            combinations.remove(0);
        }
        return combinations;
    }
}

File: answers/leetcode/Letter Combinations of a Phone Number/Letter Combinations of a Phone Number.py
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """

        combinations = []
        mapping = {
            1: "",
            2: "abc",
            3: "def",
            4: "ghi",
            5: "jkl",
            6: "mno",
            7: "pqrs",
            8: "tuv",
            9: "wxyz",
            0: " "
        }
        if len(digits) == 1:
            for i in mapping[int(digits[0])]:
                combinations.append(i)
        elif len(digits) > 1:
            temp = self.letterCombinations(digits[1 :])
            for i in temp:
                for j in mapping[int(digits[0])]:
                    combinations.append(j + i)
        return combinations
